### ISSUE TITLE
Strings: z3str3 str.substr Floating Point Exception / Segmentation Fault

### ISSUE THREAD

**Issue by marcogario** [Z3Prover/z3#1219]
I am not sure if I should report z3str3 issues here, and I apologize if this is not the right place.

On the following example using smt.string_solver=z3str3 I get a segmentation fault
```
(set-logic ALL)
(set-option :produce-models true)
(declare-const s1 String)
(declare-const index Int)
(assert (= (str.len s1) 5))
(assert (= index 2))
(assert (= ( str.++ ( str.substr s1 0 index)
                    ( str.substr s1 (+ index 1) (str.len s1)) ) s1))
(check-sat)
```

On this modified version I get Floating Point Exception
```
(set-logic ALL)
(set-option :produce-models true)
(declare-const s1 String)
(assert (= (str.len s1) 5))
(assert (= ( str.++ ( str.substr s1 0 4)
                    ( str.substr s1 5 5) ) s1))
(check-sat)
```

Z3 without z3str3 solves this correctly (UNSAT).

**Comment by NikolajBjorner** [comment]
It's good to report here for tracking and visibility. A fix is in the works, but this exposed a need for improvements that will have some wider effects.

**Comment by iwob** [comment]
It is possible that I found a very similar error, but this time without using smt.string_solver=z3str3 (with this option error also happens). With the HEAD (1a859d4591cd2ee5939d54792e0421e39a56632e) I got Floating Point Exception for this example:
```
(set-logic QF_S)
(declare-fun x () String)
(declare-fun z () String)
(assert (= x (str.substr z 1 1)))
(check-sat)
```

I noticed that marcogario used ALL logic. After changing the logic to ALL in the above script the error happens only with smt.string_solver=z3str3. Does this mean that these two logics have different default String theory solvers?
